src/components/checkboxTree/CheckboxTree.tsx

```diff
@@ -0,0 +1,145 @@
+import * as React from "react";
+
+import { HeaderIcon as ToggleIcon, SVGIcon } from "..";
+import theme from "../../theme";
+import { Checkbox } from "../checkbox/Checkbox";
+import { Flex } from "../container";
+
+import * as Styled from "./styled";
+import { CheckboxTreeProps, Composition } from "./types";
+import { getChildrenUids, getParentsUids, preProcessTree } from "./utils";
+
+export function CheckboxTree({
+  composition,
+  expanded,
+  selected,
+  onChange,
+  onExpand,
+  preferredSeperator = "->",
+}: CheckboxTreeProps) {
+  const processedTree = React.useMemo(
+    () =>
+      preProcessTree(
+        JSON.parse(JSON.stringify(composition)),
+        preferredSeperator
+      ),
+    [composition, preferredSeperator]
+  );
+
+  const [selectedList, setSelectedList] = React.useState<string[]>(
+    selected ?? []
+  );
+  const [expandedList, setExpandedList] = React.useState<string[]>(
+    expanded ?? []
+  );
+  const [indeterminate, setIndeterminate] = React.useState<string[]>([]);
+
+  React.useEffect(() => {
+    if (onChange) {
+      onChange(selectedList);
+    }
+  }, [selectedList, onChange]);
+
+  React.useEffect(() => {
+    if (onExpand) {
+      onExpand(expandedList);
+    }
+  }, [expandedList, onExpand]);
+
+  React.useEffect(() => {
+    const allUids = selectedList
+      .map((id) => getParentsUids(id, preferredSeperator))
+      .flat();
+
+    setIndeterminate(allUids);
+  }, [selectedList]);
+
+  const RenderComposition = ({
```

Comment:
feels like this could be a separate component in a separate file, what do you think?